pkg/orchestrator/data_scrubber.go

```diff
@@ -0,0 +1,147 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build orchestrator
+
+package orchestrator
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/process/config"
+)
+
+var (
+	defaultSensitiveWords = []string{
+		"password", "passwd", "mysql_pwd",
+		"access_token", "auth_token",
+		"api_key", "apikey", "pwd",
+		"secret", "credentials", "stripetoken"}
+)
+
+// DataScrubber allows the agent to blacklist cmdline arguments that match
+// a list of predefined and custom words
+type DataScrubber struct {
+	Enabled                  bool
+	StripAllArguments        bool
+	RegexSensitivePatterns   []*regexp.Regexp
+	LiteralSensitivePatterns []string
+	scrubbedCmdlines         map[string][]string
+}
+
+// NewDefaultDataScrubber creates a DataScrubber with the default behavior: enabled
+// and matching the default sensitive words
+func NewDefaultDataScrubber() *DataScrubber {
+	newDataScrubber := &DataScrubber{
+		Enabled:                  true,
+		LiteralSensitivePatterns: defaultSensitiveWords,
+		scrubbedCmdlines:         make(map[string][]string),
+	}
+
+	return newDataScrubber
+}
+func (ds *DataScrubber) ContainsBlacklistedWord(word string) bool {
+	for _, pattern := range ds.LiteralSensitivePatterns {
+		if strings.Contains(strings.ToLower(word), pattern) {
+			return true
+		}
+	}
+	return false
+}
+
+// ScrubCommand hides the argument value for any key which matches a "sensitive word" pattern.
+// It returns the updated cmdline, as well as a boolean representing whether it was scrubbed
+// future: we can add a check to do regex matching or simple matching depending whether we have RegexSensitivePatterns
+func (ds *DataScrubber) ScrubSimpleCommand(cmdline []string) ([]string, bool) {
+	changed := false
+	regexChanged := false
+	var wordReplacesIndexes []int
+
+	rawCmdline := strings.Join(cmdline, " ")
+	for _, pattern := range ds.RegexSensitivePatterns {
+		if pattern.MatchString(rawCmdline) {
+			regexChanged = true
+			rawCmdline = pattern.ReplaceAllString(rawCmdline, "${key}${delimiter}********")
+		}
+	}
+	newCmdline := strings.Split(rawCmdline, " ")
+	// preprocess, without the preprocessing it is needed to strip until the whitespaces.
+	for index, cmd := range newCmdline {
+		for _, pattern := range ds.LiteralSensitivePatterns { // this can be optimized
+			// if we found a word from the list, it means that either the current or next word should be a password we want to replace.
+			if strings.Contains(strings.ToLower(cmd), pattern) { // password=1234
```

Comment:
Not 100%, as I have not looked into it, but a suffix tree could increase the performance.